### PR TITLE
feat: multi-testnet deployments

### PR DESCRIPTION
**Pull request by prat-gpt**
## Description

Multi-testnet deployments.

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

-   [X] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
-   [ ] ~added `!` to the type prefix if API or client breaking change~
-   [ ] ~provided a link to the relevant issue or specification~
-   [X] included the necessary unit and integration tests
-   [X] code is adequately commented
-   [ ] ~updated the relevant documentation or specification~
-   [X] reviewed "Files changed" and left comments if necessary
-   [X] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

-   [x] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
-   [ ] ~confirmed `!` in the type prefix if API or client breaking change~
-   [ ] confirmed all author checklist items have been addressed
-   [ ] reviewed service design and naming
-   [ ] reviewed documentation is accurate
-   [ ] reviewed tests and test coverage
-   [ ] manually tested (if applicable)

### CODE REVIEW

File: proposals/Addresses.json
Comment (ElliotFriedman) [inline proposals/Addresses.json]: how will this work if you have multiple deployments on the same network? IMO the best approach would be to have different Addresses.json for each deployment on test networks.

File: proposals/Addresses.sol
Comment (ElliotFriedman) [inline proposals/Addresses.sol]: this check should be uncommented as there are no addresses in Addresses.json that are null.

File: proposals/zips/zip001.sol
Comment (ElliotFriedman) [inline proposals/zips/zip001.sol]: I don't feel great about this pattern of changing the address.

Would much rather see the `isAddressSet` method be used to tell if the address is set, and if not, deploy, not just deploy a new contract each time and change the address even if the address is already set.

**Comment by ElliotFriedman** [comment]
I think this makes a lot more sense to have multiple JSON files for each contract deployment on testnet. With this, you'll then need to move things to a structure where the Addresses.json file is read in from an env variable or some other way so you don't just read in a single address all of the time.

Unsure why changeAddress is being used in the zips. It makes way more sense to check if the address is set first, and then if it isn't, deploy and add the address vs just always deploying a new contract. If you always deploy a new contract, then integration tests won't work properly because you won't be testing against the mainnet state of that actual contract.

**Comment by ghost** [comment]
> I think this makes a lot more sense to have multiple JSON files for each contract deployment on testnet. With this, you'll then need to move things to a structure where the Addresses.json file is read in from an env variable or some other way so you don't just read in a single address all of the time.
> 
> Unsure why changeAddress is being used in the zips. It makes way more sense to check if the address is set first, and then if it isn't, deploy and add the address vs just always deploying a new contract. If you always deploy a new contract, then integration tests won't work properly because you won't be testing against the mainnet state of that actual contract.

Agreed. I also don't see where we are bringing in the forge proposal simulator package?

**Comment by ElliotFriedman** [comment]
> Agreed. I also don't see where we are bringing in the forge proposal simulator package?

I don't think it would make sense to bring in FPS right now, there are some pretty big changes about to land, so we should wait for them to land before bringing it in.

I think it makes the most sense to clean up the current repo, and then decide if we should bring in FPS as a submodule once things are cleaned up and sorted.

File: proposals/zips/zip003.sol
Comment (ElliotFriedman) [inline proposals/zips/zip003.sol]: All of these deployments and add addresses should use the following pattern:

1. check if address is already set using method on Addresses object `isAddressSet(NAME)`
2. if not set, deploy and add. if set, do not deploy, do not add

File: proposals/TestProposals.sol
Comment (ghost) [inline proposals/TestProposals.sol:59]: Hmm - not sure I like this. Need to think about this some more.

File: proposals/TestProposals.sol
Comment (ghost) [inline proposals/TestProposals.sol]: Why do we want this BOOTSTRAP environment variable?

**Review by ghost — CHANGES_REQUESTED** [review]

File: proposals/TestProposals.sol
Comment (intl-man) [reply proposals/TestProposals.sol:58]: I also don't think we should be checking for an explicit chain Id.

File: script/deploy/ValidProposal.s.sol
Comment (ghost) [inline script/deploy/ValidProposal.s.sol]: Shouldn't this path be dynamic?

**Review by ghost — CHANGES_REQUESTED** [review]

File: proposals/TestProposals.sol
Comment (prat-gpt) [reply proposals/TestProposals.sol:59]: Earlier the integration tests were not running against the deployed contracts. We were redeploying all the contracts and updating the addresses, so in a way we were not doing integration testing. When I ran integration tests against mainnet contracts I got to understand that we revoke the admin role from the deployer on mainnet which means for testing against mainnet we have to make the deployer as `ADMIN_MULTISIG`.
Please let me know if I am missing something here.

File: proposals/TestProposals.sol
Comment (prat-gpt) [reply proposals/TestProposals.sol]: I added this in case if we have to bootstrap a new testing environment on testnet, in this case all the zips must be simulated. If we will run the following command a new system will be bootstrapped.
`BOOTSTRAP=true forge test --fork-url $RPC_URL --match-contract TestProposals -vvv --keystore <Path> --password <Password>`
where keystore has deployer private key.

We can make a separate script for this as well. Should I add a script also?

File: script/deploy/ValidProposal.s.sol
Comment (prat-gpt) [reply script/deploy/ValidProposal.s.sol]: I assumed that we are using script only for mainnet. I can make it dynamic as well.

File: proposals/TestProposals.sol
Comment (intl-man) [reply proposals/TestProposals.sol:58]: I just don't like this. We're now hardcoding a chain ID.

File: script/deploy/ValidProposal.s.sol
Comment (ghost) [reply script/deploy/ValidProposal.s.sol]: No, it should be available for any network.

File: proposals/TestProposals.sol
Comment (intl-man) [reply proposals/TestProposals.sol:60]: Ok, but this is still incomplete because you have to go and adjust the `DeployProposal.s.sol` for running each proposal.

File: proposals/TestProposals.sol
Comment (ghost) [reply proposals/TestProposals.sol:59]: What if we have a DEPLOYER address in the address file? In `mainnet.json` it can be the `ADMIN_MULTISIG` address, and in `devnet.json` (for example) it doesn't need to change. Make sense? This way we don't need this if statement. It'll just use what's in the json file.

Comment (prat-gpt) [reply proposals/TestProposals.sol:59]: I thought of doing this but  this way we will be naming `ADMIN_MULTISIG` as `DEPLOYER` in json which is not the case. I think the current code explains the real scenario more clearly. What do u think.

File: proposals/zips/zip003.sol
Comment (prat-gpt) [reply proposals/zips/zip003.sol]: We don't need to check `isAddressSet` as the already deployed zips will only be simulated when either running tests on anvil or bootstrapping a testnet instance and in both of these cases the address will not be set for sure.

File: script/deploy/ValidProposal.s.sol
Comment (prat-gpt) [reply script/deploy/ValidProposal.s.sol]: fixed this here [4a96828](https://github.com/ZTX-Foundation/tuxedo/pull/203/commits/4a968281b6caa97c4fd60b3c4237a773ab09bc33)

File: proposals/TestProposals.sol
Comment (prat-gpt) [reply proposals/TestProposals.sol]: updated here [814fc87](https://github.com/ZTX-Foundation/tuxedo/pull/203/commits/814fc87a22c8a93aead5e38c4d33d020fc42ca73)

File: proposals/Addresses/qa.json
Comment (ElliotFriedman) [inline proposals/Addresses/qa.json:1]: would be good to see explanation for why this file is checked into git, or remove

File: proposals/TestProposals.sol
Comment (ElliotFriedman) [inline proposals/TestProposals.sol]: how do you plan to maintain this once you create zip10?

File: proposals/zips/zip001.sol
Comment (ElliotFriedman) [inline proposals/zips/zip001.sol]:
```suggestion
        addresses.addAddress("ERC1155_MAX_SUPPLY_MINTABLE_WEARABLES", address(erc1155Wearables), true);
```

no need to add block.chain id here as Addresses.sol should have a floating chain id and not anchor to the chain the contract is constructed on

File: proposals/zips/zip003.sol
Comment (ElliotFriedman) [inline proposals/zips/zip003.sol]: redundant, remove this line

Comment (ElliotFriedman) [inline proposals/zips/zip003.sol]:
```suggestion
```

**Comment by ElliotFriedman** [comment]
Would be great to see comments on how to operationalize this setup.

How to run one or all ZIP's against testnet.

How to run scripts to build and simulate calldata creation.

File: .env.example
Comment (intl-man) [inline .env.example:7]: style: add a new line to the end of the file

Comment (ghost) [inline .env.example]: style: add a new line to the end of the file

File: proposals/Addresses/localnet.json
Comment (ghost) [inline proposals/Addresses/localnet.json]: style: add a new line to the end of the file

File: proposals/Addresses/mainnet.json
Comment (ghost) [inline proposals/Addresses/mainnet.json]: style: add a new line to the end of the file

**Review by ghost — CHANGES_REQUESTED** [review]

File: proposals/Addresses/qa.json
Comment (prat-gpt) [reply proposals/Addresses/qa.json:1]: I added this file after discussion with @intl-man, that there will be multiple environments on testnet. Pushed it as it helps to understand that multiple environments are supported by us.

**Comment by prat-gpt** [comment]
> Would be great to see comments on how to operationalize this setup.
> 
> How to run one or all ZIP's against testnet.
> 
> How to run scripts to build and simulate calldata creation.

I am adding a readme in proposals folder, that would resolve these queries.

File: proposals/TestProposals.sol
Comment (prat-gpt) [reply proposals/TestProposals.sol]: Once zip10 is executed it needs to be added here. If we want to test zip10, we can replace `zipTest` with zip10.

File: script/deploy/DeployTestnet.s.sol
Comment (ElliotFriedman) [inline script/deploy/DeployTestnet.s.sol]: why is there a different function to deploy for testing? I would think it's cleaner to have a single deploy function across all proposals.

**Comment by ElliotFriedman** [comment]
This would be a good place to clean up the proposals interface. Here is what we have now:

```

    /// @notice runs before all deployments.
    /// @dev a place to put pre-deployment checks
    function _beforeDeploy(Addresses addresses, address deployer) internal virtual;

    /// @notice deployment of zip's contracts/features
    function _deploy(Addresses addresses, address deployer) internal virtual;

    /// @notice runs after all deployments
    function _afterDeploy(Addresses addresses, address deployer) internal virtual;

    /// @notice runs after all deployments for testing only
    /// @dev as an example in testing mode we dont drop the admin role as its needed for further deployments in an automated test
    function _aferDeployForTestingOnly(Addresses, address deployer) internal virtual;

    /// @notice runs after all deployments on-chain only and will ensure that admin role is revoked from deployer
    /// @dev Revoked admin role after all deployments and needs to added again before another deployment can be done.
    function _afterDeployOnChain(Addresses, address deployer) internal virtual;

    /// @notice build governance proposal
    function _build(Addresses addresses, address deployer) internal virtual;

    /// @notice run governance proposal onchain
    function _run(Addresses addresses, address deployer) internal virtual;

    /// @notice teardown anything required
    function _teardown(Addresses addresses, address deployer) internal virtual;

    /// @notice validate the deployment
    function _validate(Addresses addresses, address deployer) internal virtual;

    /// @notice validate the deployment for testing only
    /// @dev as an example in testing mode we dont drop the admin role as its needed for further deployments in an automated test
    function _validateForTestingOnly(Addresses, address deployer) internal virtual;

    /// @notice validate the deployment on-chain only
    function _validateOnChain(Addresses, address deployer) internal virtual;
```


Here is what I think we should simplify things to:

```

    /// @notice runs before all deployments.
    /// @dev a place to put pre-deployment checks
    function _beforeDeploy(Addresses addresses, address deployer) internal virtual;

    /// @notice deployment of zip's contracts/features
    function _deploy(Addresses addresses, address deployer) internal virtual;

    /// @notice runs after all deployments
    function _afterDeploy(Addresses addresses, address deployer) internal virtual;

    /// @notice build governance proposal
    function _build(Addresses addresses, address deployer) internal virtual;

    /// @notice run governance proposal onchain
    function _run(Addresses addresses, address deployer) internal virtual;

    /// @notice teardown anything required
    function _teardown(Addresses addresses, address deployer) internal virtual;

    /// @notice validate the deployment
    function _validate(Addresses addresses, address deployer) internal virtual;
```

File: proposals/TestProposals.sol
Comment (ElliotFriedman) [reply proposals/TestProposals.sol:59]: I think it's better to use incorrect naming conventions on this one address if it allows us to simplify the deploy function.

@prateek105 can we do things your suggested way and remove all the different test functions within proposals? If so, that sounds good. If not, I would make the names uniform if that can fix this issue.

File: script/deploy/DeployTestnet.s.sol
Comment (prat-gpt) [reply script/deploy/DeployTestnet.s.sol]: The system was already built like this. This was done because the admin role is revoked from deployer on mainnet in zip001 [here](https://github.com/ZTX-Foundation/tuxedo/blob/6530048b0400d80563489671319e184610fdbd3e/proposals/zips/zip001.sol#L67) while that is not done on testnets.

I am working on making it a single contract in a separate PR.

**Comment by prat-gpt** [comment]
> Here is what I think we should simplify things to:
> 
> ```
> 
>     /// @notice runs before all deployments.
>     /// @dev a place to put pre-deployment checks
>     function _beforeDeploy(Addresses addresses, address deployer) internal virtual;
> 
>     /// @notice deployment of zip's contracts/features
>     function _deploy(Addresses addresses, address deployer) internal virtual;
> 
>     /// @notice runs after all deployments
>     function _afterDeploy(Addresses addresses, address deployer) internal virtual;
> 
>     /// @notice build governance proposal
>     function _build(Addresses addresses, address deployer) internal virtual;
> 
>     /// @notice run governance proposal onchain
>     function _run(Addresses addresses, address deployer) internal virtual;
> 
>     /// @notice teardown anything required
>     function _teardown(Addresses addresses, address deployer) internal virtual;
> 
>     /// @notice validate the deployment
>     function _validate(Addresses addresses, address deployer) internal virtual;
> ```
I am working on exactly this in a separate PR as I think making that change in this PR will make it heavy. We can merge this one if everything else looks good and I will make a separate PR for this. what do u think?

**Comment by ElliotFriedman** [comment]
> I am working on exactly this in a separate PR as I think making that change in this PR will make it heavy. We can merge this one if everything else looks good and I will make a separate PR for this. what do u think?

That sounds good!

File: proposals/Addresses/qa.json
Comment (ghost) [reply proposals/Addresses/qa.json:1]: @prateek105 lets remove it for now. We can populate it once we have done a full deployment for QA.

File: proposals/TestProposals.sol
Comment (intl-man) [reply proposals/TestProposals.sol:72]: Again, why do we lock this down to a local chain Id? Maybe I'm missing something?

File: script/deploy/DeployTestnet.s.sol
Comment (intl-man) [reply script/deploy/DeployTestnet.s.sol:63]: Can't we just update this PR instead?

File: proposals/README.md
Comment (ghost) [inline proposals/README.md]: Do we really need to include this? Could we link to an example file instead?

File: proposals/README.md
Comment (ghost) [inline proposals/README.md]: Use NPM to run these commands. Take a look at `package.json`, as that will need some cleaning up too.

File: proposals/README.md
Comment (ghost) [inline proposals/README.md]: Also move to NPM.

File: proposals/README.md
Comment (ghost) [inline proposals/README.md]: Just link to an example file, as per my comment about the contract code above.

File: proposals/README.md
Comment (ghost) [inline proposals/README.md]: Again, move to NPM.

**Review by ghost — CHANGES_REQUESTED** [review]

File: proposals/README.md
Comment (ghost) [reply proposals/README.md]: Could, for example, live in a `docs/examples/proposal.sol` file.

File: proposals/README.md
Comment (ghost) [reply proposals/README.md]: Could, for example, live in a `docs/examples/network.json` file.

File: proposals/README.md
Comment (prat-gpt) [reply proposals/README.md]: We don't need this. removed it.

**Review by ghost — APPROVED** [review]

**Review by ElliotFriedman — APPROVED** [review]
LGTM!

**Pull request merged**
